\documentclass[12pt]{article}
\usepackage[margin=0.5in]{geometry}
\usepackage{mathptmx}
\usepackage{amsmath}
\usepackage{tikz}
\usepackage{parskip}
\usepackage{mathtools}
\usepackage{multirow}
\pagestyle{empty}

\usetikzlibrary{positioning}

\newcommand{\ccf}{\mathrm{CCF}}
\newcommand{\vaf}{\mathrm{VAF}}
\newcommand{\prev}{\mathrm{prevalence}}
\newcommand{\pur}{\mathrm{tumour\;purity}}
\newcommand{\cn}{\mathrm{copy\;number}}
\newcommand{\macn}{\mathrm{major\;CN}}
\newcommand{\micn}{\mathrm{minor\;CN}}

\begin{document}

\section{Assumption}

If a mutation is present on the amplified strand of a CNA amplification event,
then it is present in the \emph{entire} subclone containing that event.
Presence on a duplicated strand implies the mutation was present before the
duplication happened, that is, the cells with the mutation are a superset of
the cells with the CNA.

\section{Terminology}

\begin{description}
    \item[VAF] variant allele fraction, ie. the proportion of reads supporting a mutation
    \item[CCF] cancer cell fraction, ie. the proportion of tumour cells containing a mutation
    \item[prevalence] the proportion of tumour cells carrying a particular CNA
    \item[purity] the proportion of cells in the sample deriving from the tumour
    \item[copy number] the number of copies of a locus carried by the tumour
        cells with a CNA
    \item[major CN] copy number of strand with greater number of copies
    \item[minor CN] copy number of strand with lesser number of copies
\end{description}

Note that major CN + minor CN = copy number. Also, we refer to the strand with
the major copy number as the major strand, and likewise for the minor strand.

\section{Correction}

The average number of copies of a locus in a tumour cell is

\begin{align*}
    \cn \cdot \prev + 2 \cdot (1-\prev)
\end{align*}

The average number of copies in a normal cell is 2. From this, we can calculate
the average number of copies of a locus in a cell from a sample of mixed tumour
and normal cells. Since we will use this quantity often, we will call it
$\alpha$. That is,

\begin{align*}
    \alpha = [(\cn \cdot \prev) + (2 \cdot (1-\prev))] 
             \cdot \pur + 2 \cdot (1 - \pur)
\end{align*}

The VAF is the average number of copies carrying the mutation, divided by the
average number of total copies of the locus, in a cell in the sample. 

Suppose the mutation occurs on a strand without a dupllication (that is, the
strand has either zero or one copies the locus in question). Then each tumor
cell with the mutation has exactly one copy of the mutation, so the average
number of copies of the mutated allele per cell is

\begin{align*}
    \pur \cdot \ccf,
\end{align*}

and hence

\begin{align*}
    \vaf = \frac{\pur \cdot \ccf}{\alpha} \quad\Rightarrow\quad
    \ccf = \frac{\alpha \cdot \vaf}{\pur}.
\end{align*}

This is also the correct formula to use if the mutation does not occur in cells
with the CNA.

Now, suppose there is a CNA at the locus of the mutation in the tumour.
Consider the case where the mutation is on the major strand, that is, there are
major CN copies of the mutation in cells with the CNA. By assumption, in this
scenario, the population of cells showing the duplication is a subset of the
population with the mutation, which is in turn a subset of the complete
population of tumour cells. The below schematic illustrates this.

\begin{center}
    \begin{tikzpicture}
        \node [circle, draw, inner sep=2cm] (a) { };
        \node [below=0.3cm of a.north, anchor=north] {neither};
        \node [circle, draw, above=0cm of a.south, inner sep=1.7cm, anchor=south, fill=blue, fill opacity=0.25] (b) { };
        \node [below=0.3cm of b.north, anchor=north] {mutation};
        \node [circle, draw, above=0cm of a.south, inner sep=1.2cm, anchor=south, fill=blue, fill opacity=0.25] (c) { };
        \node [below=0cm of c.center, anchor=center] {CNA};

        \node [right=of a, text width=7cm] {
            CCF = area of light blue circle \\
            prevalence = area of dark blue circle
        };
    \end{tikzpicture}
\end{center}

The cells without the mutation (white region) do not have the mutation at all;
in other words, there are zero copies of the mutation in these cells. The cells
without the CNA (the light blue region) do not have a duplication, so they
possess exactly one copy of the mutation each. In the cells with the CNA (dark
blue region), there are major CN copies of the mutation. Hence, the average
number of copies of the mutation in a tumour cell is

\begin{align*}
     & 0 \cdot (1-\ccf) \\
    +& 1 \cdot (\ccf - \prev) \\
    +& \macn \cdot \prev,
\end{align*}

and so

\begin{align*}
    \vaf &= \frac{[\ccf - \prev + \macn \cdot \prev] \cdot \pur}{\alpha} \\
    \Rightarrow \ccf &= \frac{\alpha \vaf}{\pur} - (\macn - 1) \cdot \prev.
\end{align*}

This formula for CCF is intuitive. Since the mutation is duplicated, there are
$[(\macn - 1)\cdot \prev]$ ``extra'' copies of the mutation which must be
subtracted off to get the true CCF.

Putting this all together, we arrive at an expression for CCF for a mutation on
the major strand.

\begin{align*}
    \ccf =
    \begin{dcases}
        \frac{\alpha\cdot\vaf}{\pur} & \macn < 2 \\
        \frac{\alpha\cdot\vaf}{\pur} - (\macn - 1) \cdot \prev & \macn \geq 2.
    \end{dcases}
\end{align*}

Equivalently,

\begin{align*}
    \ccf = \frac{\alpha\cdot\vaf}{\pur} - \max(\macn - 1, 0) \cdot \prev.
\end{align*}

Of course, the expression for CCF when the mutation is on the minor strand is
exactly the same, except replacing ``$\micn$'' with ``$\macn$''. Therefore, 

\begin{align}
    \ccf = 
    \begin{dcases}
        \frac{\alpha\cdot\vaf}{\pur} 
        & \text{mutation not on CNA} \\
        \frac{\alpha\cdot\vaf}{\pur} - \max(\micn - 1, 0) \cdot \prev
        & \text{mutation on minor strand} \\
        \frac{\alpha\cdot\vaf}{\pur} - \max(\macn - 1, 0) \cdot \prev
        & \text{mutation on major strand}.
    \end{dcases}
    \label{eq:cor}
\end{align}

\section{Choosing a scenario}

Unfortunately, we don't know \textit{a priori} whether a mutation co-occurs
with the CNA, and if so, whether it is on the major or minor strand. But we can
make some observations, which depend on the major and minor copy numbers.

\subsection{No duplications}

The simplest case is when neither strand is duplicated, that is, $\macn \leq 1$
and $\micn \leq 1$. In this case, the mutation must occur at most once in each
cell, so

\begin{align*}
    \ccf = \frac{\alpha\cdot\vaf}{\pur}.
\end{align*}

\subsection{One duplication without LOH}

This case occurs when the major strand is duplicated ($\macn > 1$), but the
minor strand is left unchanged ($\micn = 1$). We have

\begin{align*}
    \ccf = 
    \begin{dcases}
        \frac{\alpha\cdot\vaf}{\pur} - (\macn - 1) \cdot \prev
        & \text{mutation on major strand} \\
        \frac{\alpha\cdot\vaf}{\pur} 
        & \text{otherwise}.
    \end{dcases}
\end{align*}

If the mutation is on the duplicated strand, the CCF must fall between $\prev$
and 1, by assumption. Otherwise, the CCF must be between 0 and 1. To translate
these bounds into VAF values, we substitute into the equation above. The
resulting VAF intervals may overlap, resulting in ``ambiguous'' VAFs, or may
have a gap between them, containing ``impossible'' VAFs. To resolve these
cases, we use the midpoint of the problematic interval (ie. the overlap or the
gap) as the separator between the two cases. This midpoint is

\begin{align*}
    \vaf = \frac{(\macn \cdot \prev + 1) \cdot \pur}{2\alpha},
\end{align*}

and therefore,

\begin{align*}
    \ccf = 
    \begin{dcases}
        \frac{\alpha\cdot\vaf}{\pur} - (\macn - 1) \cdot \prev
        & \vaf \geq \frac{(\macn \cdot \prev + 1) \cdot \pur}{2\alpha} \\
        \frac{\alpha\cdot\vaf}{\pur} 
        & \text{otherwise}.
    \end{dcases}
\end{align*}

\subsection{One duplication with LOH}

This case occurs when the major strand is duplicated ($\macn > 1$) while the
minor strand is deleted ($\micn = 0$). The relationship between CCF and VAF is
the same as in the previous case, but the allowed CCF values are different.
Since the minor strand has no copies in the CNA region, if the mutation is not
on the major strand, then it must have CCF no more than $1-\prev$. The new
midpoint of the allowed regions is

\begin{align*}
    \vaf = \frac{[1 + (\macn - 1) \cdot \prev] \cdot \pur}{2\alpha},
\end{align*}

so in this case,

\begin{align*}
    \ccf = 
    \begin{dcases}
        \frac{\alpha\cdot\vaf}{\pur} - (\macn - 1) \cdot \prev
        & \vaf \geq \frac{[1 + (\macn - 1) \cdot \prev] \cdot \pur}{2\alpha} \\
        \frac{\alpha\cdot\vaf}{\pur} 
        & \text{otherwise}.
    \end{dcases}
\end{align*}

\subsection{Two duplications}

The final case occurs when both the major and minor strand are duplicated
($\macn > 1$ and $\micn > 1$). There are three possibilities: the mutation may
be on the major strand, on the minor strand, or may not co-occur with the CNA
at all. Hence,

\begin{align*}
    \ccf = 
    \begin{dcases}
        \frac{\alpha\cdot\vaf}{\pur} 
        & \text{mutation not on CNA} \\
        \frac{\alpha\cdot\vaf}{\pur} - (\micn - 1) \cdot \prev
        & \text{mutation on minor strand} \\
        \frac{\alpha\cdot\vaf}{\pur} - (\macn - 1) \cdot \prev
        & \text{mutation on major strand}.
    \end{dcases}
\end{align*}

The mutation is allowed to be not on the CNA when the CCF would fall in $[0,
1-\prev]$. In the other two cases, the CCF must fall in $[\prev, 1]$.
Substituting these CCF values into the above, we get one allowed VAF interval
for each case. Again, we split between the cases using the midpoint of the
bounds of the adjacent intervals. For brevity, we shall call these $\beta$ and
$\gamma$.

\begin{align*}
    \beta &= \frac{[1 + (\micn - 1) \cdot \prev] \cdot \pur}{2\alpha} \\
    \gamma &= \frac{[1 + (\macn + \micn - 1) \cdot \prev] \cdot \pur}{2\alpha}.
\end{align*}

Therefore,

\begin{align*}
    \ccf = 
    \begin{dcases}
        \frac{\alpha\cdot\vaf}{\pur} 
        & \vaf < \beta \\
        \frac{\alpha\cdot\vaf}{\pur} - (\micn - 1) \cdot \prev
        & \beta \leq \vaf < \gamma \\
        \frac{\alpha\cdot\vaf}{\pur} - (\macn - 1) \cdot \prev
        & \vaf \geq \gamma.
    \end{dcases}
\end{align*}

\section{Example}

We illustrate this correction below for tumour purity = 0.8 and prevalence =
0.5. The blue, green, and red lines correspond to the relationships between VAF
and CCF when, respectively, the mutation is not on the CNA, when it is on the
minor strand, and when it is on the major strand. The thick regions are where
the scenario is allowed to be used, and the black line shows the correction
used. The lines are offset slightly for visibility.

\begin{center}
<<echo=F, eval=T, dev='pdf', fig.height=5>>=
par(mar=c(4,4,4,0), mfrow=c(2,3))
sep <- 0.04
vaf <- seq(0, 1, 0.01)
pur <- 0.8
prev <- 0.5

vaf.to.ccf <- function (vaf, pur, prev, minor.cn, major.cn) {
    cn <- major.cn + minor.cn
    alpha <- (cn*prev + 2*(1-prev))*pur + 2*(1-pur)

    if (minor.cn <= 1 & major.cn <= 1) {
        alpha*vaf/pur

    } else if (minor.cn == 1) {
        if (vaf >= (major.cn*prev+1)*pur/(2*alpha))
            alpha*vaf/pur - (major.cn-1)*prev
        else
            alpha*vaf/pur

    } else if (minor.cn == 0) {
        if (vaf >= (1+(major.cn-1)*prev)*pur/(2*alpha))
            alpha*vaf/pur - (major.cn-1)*prev
        else
            alpha*vaf/pur

    } else {
        if (vaf <= (1+(minor.cn-1)*prev)*pur/(2*alpha))
            alpha*vaf/pur
        else if (vaf >= (1+(major.cn+minor.cn-1)*prev)*pur/(2*alpha))
            alpha*vaf/pur - (major.cn-1)*prev
        else
            alpha*vaf/pur - (minor.cn-1)*prev
    }
}

do.plot <- function (major.cn, minor.cn) {
    cn <- major.cn + minor.cn
    alpha <- (cn*prev + 2*(1-prev))*pur + 2*(1-pur)
    
    ccf.no.cna <- alpha*vaf/pur
    ccf.minor <- alpha*vaf/pur - max(minor.cn-1, 0)*prev 
    ccf.major <- alpha*vaf/pur - max(major.cn-1, 0)*prev
    ccf.real <- mapply(vaf.to.ccf, vaf, pur, prev, minor.cn, major.cn)
    
    no.cna.ok <- 0 <= ccf.no.cna & 1-prev >= ccf.no.cna
    minor.ok <- prev <= ccf.minor & 1 >= ccf.minor
    major.ok <- prev <= ccf.major & 1 >= ccf.major
    
    plot(vaf, ccf.real - sep, lwd=3, xlab='VAF', ylab='CCF', type='l',
         main=paste('major CN =', major.cn, ', minor.cn =', minor.cn))
    
    lines(vaf, ccf.no.cna, col='blue') 
    if (minor.cn > 0)
        lines(vaf, ccf.minor + sep, col='forestgreen')
    lines(vaf, ccf.major + 2*sep, col='red')
    
    lines(vaf[no.cna.ok], ccf.no.cna[no.cna.ok], col='blue', lwd=3)
    if (minor.cn > 0)
        lines(vaf[minor.ok], ccf.minor[minor.ok] + sep, col='forestgreen', lwd=3)
    lines(vaf[major.ok], ccf.major[major.ok] + 2*sep, col='red', lwd=3)

    if (minor.cn > 0)
        legend('bottomright', 
               legend=c('no CNA', 'minor', 'major'),
               col=c('blue', 'forestgreen', 'red'),
               lty=1, bg='white')
    else
        legend('bottomright', 
               legend=c('no CNA', 'major'),
               col=c('blue', 'red'),
               lty=1, bg='white')
}

do.plot(0, 0)
do.plot(1, 0)
do.plot(2, 0)
do.plot(1, 1)
do.plot(2, 1)
do.plot(3, 2)
@
\end{center}

\newpage

\section{Code}

The R function to do this correction is as follows.

\begin{verbatim}
vaf.to.ccf <- function (vaf, pur, prev, minor.cn, major.cn) {
    cn <- major.cn + minor.cn
    alpha <- (cn*prev + 2*(1-prev))*pur + 2*(1-pur)

    # no duplications
    if (minor.cn <= 1 & major.cn <= 1) {
        alpha*vaf/pur

    # one duplication, no LOH
    } else if (minor.cn == 1) {
        if (vaf >= (major.cn*prev+1)*pur/(2*alpha))
            alpha*vaf/pur - (major.cn-1)*prev
        else
            alpha*vaf/pur

    # one duplication with LOH
    } else if (minor.cn == 0) {
        if (vaf >= (1+(major.cn-1)*prev)*pur/(2*alpha))
            alpha*vaf/pur - (major.cn-1)*prev
        else
            alpha*vaf/pur

    # two duplications
    } else {
        if (vaf <= (1+(minor.cn-1)*prev)*pur/(2*alpha))
            alpha*vaf/pur
        else if (vaf >= (1+(major.cn+minor.cn-1)*prev)*pur/(2*alpha))
            alpha*vaf/pur - (major.cn-1)*prev
        else
            alpha*vaf/pur - (minor.cn-1)*prev
    }
}
\end{verbatim}
\end{document}
